sample/trace_key.go

```diff
@@ -112,3 +108,92 @@ outer:
 
 	return key.String(), fieldCount
 }
+
+// distinctValue keeps track of distinct values for a set of fields.
+// It stores the unique values as strings.
+type distinctValue struct {
+	buf    []byte
+	fields []string
+	values []map[uint64]string
+
+	// totalUniqueCount keeps track of how many unique values we've seen so far for a trace key.
+	totalUniqueCount int
+	// maxDistinctValue is the maximum number of distinct values we will store for a trace key.
+	maxDistinctValue int
+}
+
+func newDistinctValue(fields []string, maxDistinctValue int) *distinctValue {
+	d := &distinctValue{
+		buf:              make([]byte, 0, 1024),
+		fields:           fields,
+		values:           make([]map[uint64]string, len(fields)),
+		maxDistinctValue: maxDistinctValue,
+	}
+
+	for i := range d.values {
+		d.values[i] = make(map[uint64]string, maxDistinctValue)
+	}
+
+	return d
+
+}
+
+// Values returns the distinct values for a given field index.
+// It returns a sorted slice of strings containing the unique values for that field.
+func (d *distinctValue) Values(fieldIdx int) []string {
+	if fieldIdx < 0 || fieldIdx >= len(d.values) {
+		return nil
+	}
+
+	// Get the map for the specified field index
+	valueMap := d.values[fieldIdx]
+	if len(valueMap) == 0 {
+		return nil
+	}
+
+	values := slices.Collect(maps.Values(valueMap))
+	sort.Strings(values)
+
+	return values
+}
+
+// AddAsString adds a value to the distinct values for a given field index.
+// It returns true if the value was added, false if it was already present or if the maxDistinctValue limit was reached.
+func (d *distinctValue) AddAsString(value any, fieldIdx int) bool {
+	if value == nil {
+		return false
+	}
+
+	d.buf = d.buf[:0] // reset the buffer for each new value
+
+	switch v := value.(type) {
+	case string:
+		d.buf = append(d.buf, []byte(v)...)
+	case int:
+		d.buf = binary.BigEndian.AppendUint64(d.buf, uint64(v))
+	case int64:
+		d.buf = binary.BigEndian.AppendUint64(d.buf, uint64(v))
+	case float64:
+		d.buf = binary.BigEndian.AppendUint64(d.buf, math.Float64bits(v))
+	case bool:
+		if v {
+			d.buf = append(d.buf, '1')
+		} else {
+			d.buf = append(d.buf, '0')
+		}
+	default:
+		d.buf = append(d.buf, fmt.Sprintf("%v", v)...)
+	}
+
+	hash := metro.Hash64(d.buf, 0)
+	if _, exists := d.values[fieldIdx][hash]; !exists {
+		d.totalUniqueCount++
+		if d.totalUniqueCount >= d.maxDistinctValue {
+			return false
+		}
+		d.values[fieldIdx][hash] = fmt.Sprintf("%v", value)
```

Comment:
If we're trying to avoid calling `Sprintf`, is there a way we can use what we learned about `value` type fromt he switch statement above to avoid this call whenever possible?